src/gf.c

```diff
@@ -2037,14 +2042,20 @@ JL_DLLEXPORT void jl_method_table_add_backedge(jl_methtable_t *mt, jl_value_t *t
         else {
             // check if the edge is already present and avoid adding a duplicate
             size_t i, l = jl_array_nrows(mt->backedges);
+#ifndef JL_NDEBUG
+            // It is the caller's (inference's) responsibility to de-duplicate edges. Here we are only
+            // checking its work.
+            int found = 0;
             for (i = 1; i < l; i += 2) {
                 if (jl_array_ptr_ref(mt->backedges, i) == (jl_value_t*)caller) {
                     if (jl_types_equal(jl_array_ptr_ref(mt->backedges, i - 1), typ)) {
-                        JL_UNLOCK(&mt->writelock);
-                        return;
+                        found = 1;
+                        break;
                     }
                 }
             }
+            assert(!found && "duplicate back-edge registered");
+#endif
             // reuse an already cached instance of this type, if possible
             // TODO: use jl_cache_type_(tt) like cache_method does, instead of this linear scan?
```

Comment:
It's worth seeing if we can bypass this remaining linear scan in the MethodTable edge insertion.

If I disable it manually, the timing drops to 440 milliseconds